Update 0.6.0

AngelHeal opened the pull request:
Hi, check if this improvement is going to happen, if everything is fine, make an update to the plugin.

AngelHeal committed:
Big Update 0.6.0

AngelHeal committed:
Fix text

goldenstein64 commented:
I just saw this, I can tell you put a lot of work into it. I want to try playtesting it first before just releasing it.

I noticed you made pushes chargeable now, is there an option for disabling it? It might help with making pushes more snappy by activating it on key press rather than release. I'm not sure how much I can help with implementing this, I'm not really a Unity developer, just a C# one.

Also, there was this "buggy" behavior that I wanted to make a feature, which is pushing _yourself_. My friends and I really enjoy this part of the mod because it interacts with other mods (like HardcoreParkour) in a really funny way, and that's kind of the only reason I was messing with it (besides just making it available for the Mesa update).

AngelHeal commented:
When pressing SHIFT, for some reason, the raycast saw the local character as a character to whom force needed to be applied, which is why there was a glitchy operation.

By simply pressing F for at least 0.0 seconds, a push is triggered anyway, but with less force than charging it, it works fast enough, although tests are still needed, I called my friends, but literally for 5 minutes to test.

I can make a configuration in the config to disable charging and use the standard version.

Problems with Bing-Bong, in a regular plugin it gets the name of the item and it is different in different languages, so there it is determined by the ItemTags of item.

There were also problems using the raycast, when I picked up the bing bong, the raycast rested on it and could not capture the player, so I had to use it only in the LayerMask "Character" so that it skipped everything except Character.

AngelHeal commented:
"I noticed you made pushes chargeable now, is there an option for disabling it?"
Only at the same time, the stamina is not spent, since in my implementation it depends on currentCharge, but this can be corrected.
<img width="636" height="325" alt="image" src="https://github.com/user-attachments/assets/95ce72bb-7c8d-4c20-990d-c0c01585e92d" />

goldenstein64 committed:
fix: add animationTime back

goldenstein64 committed:
style: run formatter

goldenstein64 committed:
refactor: remove animationTime after discovering redundant ANIMATION_TIME

goldenstein64 committed:
refactor: PlayPushAnimation can accept a null Character

goldenstein64 committed:
perf: use pattern matching on null and item tag comparison

goldenstein64 committed:
refactor: rename some variables in MakeTex

AngelHeal committed:
Update Plugin.cs

Fixed ANIMATION_TIME and added canPush to settings in order to disable push charging

AngelHeal committed:
Update ConfigurationHandler.cs

canCharge added

AngelHeal commented:
Fixed ANIMATION_TIME and added canCharge* to settings in order to disable push charging.

Can I add an additional button that will determine if the character is pushing himself?
If you just press F, will it push everything except itself, if you hold down ALT + F will it push itself?
It's just that in the usual case, you can make a mistake and push yourself into the abyss.
Waiting for your feedback.

AngelHeal committed:
Merge branch 'main' into main

AngelHeal committed:
fix UseStamina without CanCharge

goldenstein64 committed:
Merge branch 'main' of https://github.com/AngelHeal/PEAK-PushMod-MesaUpdate into AngelHeal/main

goldenstein64 committed:
refactor: minor edits

goldenstein64 commented on src/PushMod_MesaUpdate/ConfigurationHandler.cs:
```diff
@@ -0,0 +1,57 @@
+    private ConfigEntry<bool> _configcanCharge;
+
+    public KeyCode PushKey => ParseKeyCode(_configPushKey.Value, KeyCode.F);
+    public KeyCode ProtectionKey => ParseKeyCode(_configProtectionKey.Value, KeyCode.F11);
```
This (line 17) is different from line 32, which says F10

goldenstein64 commented:
I noticed that the config menu for this mod is missing, I am not sure what caused that to happen. I _also_ don't know how these config options are supposed to be accessed other than through the mod settings menu, I have ModConfig and SettingsExtenderForked installed.

Edit: Uh, I forgot to try editing thunderstore's .cfg files, oops!

AngelHeal committed:
edit configProtectionKey

AngelHeal committed:
Fixed the display in ModConfig, now the config is created automatically instead of manually via ConfigurationHandler

AngelHeal commented:
> I noticed that the config menu for this mod is missing, I am not sure what caused that to happen. I _also_ don't know how these config options are supposed to be accessed other than through the mod settings menu, I have ModConfig and SettingsExtenderForked installed.
> 
> Edit: Uh, I forgot to try editing thunderstore's .cfg files, oops!

Fixed it, for some reason, if you use manual config creation, ModConfig does not see this mod.
<img width="1340" height="402" alt="image" src="https://github.com/user-attachments/assets/4e47fad9-37ba-494e-b7fe-88a24b26c386" />

goldenstein64 committed:
feat: store config keys as keycodes

goldenstein64 commented:
> Fixed ANIMATION_TIME and added canCharge* to settings in order to disable push charging.
> 
> Can I add an additional button that will determine if the character is pushing himself? If you just press F, will it push everything except itself, if you hold down ALT + F will it push itself? It's just that in the usual case, you can make a mistake and push yourself into the abyss. Waiting for your feedback.

I'm sorry for getting back to you so late on this. This sounds like a great idea, although Alt + F may not be the best binding since the spacebar can't be pressed at the same time. Just having a separate key for self-pushing would be good enough, defaulting to something like G.

goldenstein64 committed:
feat: lower max charge to 1 second

goldenstein64 committed:
feat: add SelfPushKey

goldenstein64 committed:
feat: make SelfPushKey work

AngelHeal commented on src/PushMod_MesaUpdate/Plugin.cs:
```diff
@@ -39,7 +39,7 @@ public class PushManager : MonoBehaviour {
     private const float STAMINA_COST = 0.1f;                    // Stamina consumed per push
 
-    private const float MAX_CHARGE = 3f;                        // Maximum charge duration (seconds)
+    private const float MAX_CHARGE = 1f;                        // Maximum charge duration (seconds)
```
Then the meaning of Mathf.Max(currentCharge, 1f) line 188 is lost.

goldenstein64 replied:
The previous behavior could be replicated by replacing line 188's `currentCharge` with `3 * currentCharge / MAX_CHARGE`. It would definitely feel arbitrary though, adding a hidden constant in there, I would opt for a `MAX_STAMINA_COST_MULTIPLIER` constant.

goldenstein64 commented:
So far, my friends and I think this is the best version of the mod so far, as of the commit I just pushed. Let me know if there are any changes you are unhappy with or just any suggestions.

LGTM :tm:

goldenstein64 commented on src/PushMod_MesaUpdate/Plugin.cs:
```diff
@@ -1,159 +1,390 @@
+        // Calculate final push force with multipliers
+        float chargeMultiplier = 1f + (currentCharge / MAX_CHARGE) * CHARGE_FORCE_MULTIPLIER;
+        float bingBongMultiplier = bingBong ? BINGBONG_MULTIPLIER : 1f;
+        float totalMultiplier = bingBongMultiplier + chargeMultiplier;
```
I think this line may be wrong given the bing-bong multiplier is 1 when absent. Maybe change it to this?

```suggestion
        float totalMultiplier = bingBongMultiplier * chargeMultiplier;
```

AngelHeal committed:
Update src/PushMod_MesaUpdate/Plugin.cs

Co-authored-by: goldenstein64 <goldnstein64@gmail.com>

AngelHeal committed:
I removed the TryPushSelf method and combined the logic into TryPush, why do we need 2 identical methods?)

AngelHeal commented:
I think we did a good job, it remains to solve the MAX_CHARGE and the mod is ready)

AngelHeal committed:
Whoops, why does everyone need to know that I want to touch myself)

AngelHeal committed:
Revert "I removed the TryPushSelf method and combined the logic into TryPush, why do we need 2 identical methods?)"

This reverts commit 72c356bb176ad992245e9aa5fd630948cd4b149f.

AngelHeal committed:
Revert "Whoops, why does everyone need to know that I want to touch myself)"

This reverts commit 2bda33677ff629ce9dcc806a507e319c3903c335.

AngelHeal committed:
I removed the TryPushSelf method and combined the logic, why do we need 2 identical methods?)

AngelHeal commented on src/PushMod_MesaUpdate/Plugin.cs:
```diff
@@ -1,159 +1,374 @@
+        Plugin.Log.LogInfo($"Push force direction: {forceDirection}");
+
+        // Trigger jump SFX on the target (temporary feedback)
+        PlayPushSFX(pushedCharacter);
```
Need to test TryPushTarget with a local player when sending RPC
If the player is local, PlayPushSFX is called 2 times
in TryPushTarget and PushPlayer_Rpc

goldenstein64 replied:
I playtested with my friends looking at this, and we couldn't tell the difference in volume, assuming playing twice = twice as loud because of layering

AngelHeal committed:
The RPC has been changed, now the push animation is synchronized with all players.

With a local push, one SFX is played.

Added MAX_STAMINA_COST_MULTIPLIER, redesigned the use of stamina

AngelHeal committed:
Update CHANGELOG.md

AngelHeal commented:
@goldenstein64 can release it?
If it's not difficult for you, you can check everything related to the text, since English is not my language.

AngelHeal committed:
PlayPushSFX Remove test

goldenstein64 commented:
I will release it, yes.

Overall, thank you for improving the mod, this was a great help.
